haskell/src/Lorentz/Contracts/Stablecoin.hs

```diff
@@ -347,3 +346,46 @@ deriving anyclass instance IsoValue (MetadataRegistryStorageView)
 
 mkMetadataRegistryStorage :: big_map FA2.TokenId FA2.TokenMetadata -> MetadataRegistryStorage' big_map
 mkMetadataRegistryStorage bm = MetadataRegistryStorage () bm
+
+
+-- This TemplateHaskell splice is here to ensure the Michelson representation
+-- (i.e., the generated tree of `or` and `pair`) of these Haskell data
+-- types matches exactly the Michelson representation of the Ligo data types.
+--
+-- For example, if a Ligo data type generates a Michelson balanced tuple like
+-- ((a, b), (c, d)), but the corresponding Haskell data type generates a Michelson
+-- tuple like (a, (b, (c, d))), compilation should fail.
+--
+-- The reason they need to match is because of the way permits work.
+-- If we issue a permit for an entrypoint and the tree of `or`s is incorrect,
+-- then the permit will be unusable.
+-- See TZIP-17 for more info.
+--
+-- The splice attempts to parse the stablecoin.tz contract at compile-time.
+-- If, for example, the Michelson representation of Haskell's
+-- `Lorentz.Contracts.Stablecoin.Parameter` is different from Ligo's `parameter`,
+-- then this splice will raise a compilation error.
+--
+-- This can usually be fixed by writing a custom instance of Generic using `customGeneric`,
+-- and making sure the data type's layout matches the layout in stablecoin.tz.
+-- See examples above.
+$(case readContract @(ToT Parameter) @(ToT Storage) stablecoinPath $(embedStringFile stablecoinPath) of
+    Left e -> fail (pretty e)
+    Right _ -> pure []
+ )
+
+-- | Parse the stablecoin contract.
+parseStablecoinContract :: MonadThrow m => m (T.Contract (ToT Parameter) (ToT Storage))
```

Comment:
If we have that splice above, we know in compile time that the contract is valid (otherwise compilation fails). Can we use this knowledge here to avoid `MonadThrow` and change it to just `T.Contract (ToT Parameter) (ToT Storage)`? I am bad at TH, but I suspect it should be straightforward.

Reply:
I don't _think_ we can, or at least not in that way.

The TH splice would have to compile the contract (already done), and then somehow emit a valid [haskell expression](https://hackage.haskell.org/package/template-haskell-2.16.0.0/docs/Language-Haskell-TH.html#t:Exp) of type `T.Contract (ToT Parameter) (ToT Storage)`, as if it had been handwritten (not easily done).

In other words, what would the code generated by TH look like?

```
stablecoinContract :: T.Contract (ToT Parameter) (ToT Storage)
stablecoinContract = ???
```

Theoretically, the TH splice could pattern match on the parsed `T.Contract (ToT Parameter) (ToT Storage)`, and build an `Exp` AST representing calls to `Michelson.Typed.Instr` and `Michelson.Typed.Value`. It is doable, but it's a lot of work for little gain.

However, because the TH splice proves the contract is well-typed, we can change `either throwM` to `either error`, and get rid of `MonadThrow`, and be sure that `error` will never be called.

Reply:
Okay

> However, because the TH splice proves the contract is well-typed, we can change `either throwM` to `either error`, and get rid of MonadThrow, and be sure that error will never be called.

I think it's a good idea (as long as there is a comment explaining why the error can't happen here). It matches the idea of using `error` for errors that can happen only in case of a programmer mistake.

Reply:
Done, looks a lot better now!

Reply:
:cool: